sway-core/src/control_flow_analysis/analyze_return_paths.rs

```diff
@@ -133,7 +135,7 @@ impl<'cfg> ControlFlowGraph<'cfg> {
 /// The resulting edges from connecting a node to the graph.
 enum NodeConnection {
     /// This represents a node that steps on to the next node.
-    NextStep(Vec<NodeIndex>),
+    NextStep(Option<NodeIndex>),
```

Comment:
I'm not quite sure I understand from the code why this needs to become an Option, can you explain that?

Reply:
Two cases where we need it:

- Error nodes:
```
190:        ty::TyAstNodeContent::Error(_, _) => Ok(NodeConnection::NextStep(None)),
```

Looking at the code I notice we're not being entirely consistent wrt. how we handle error nodes - in some places we return `NextStep(None)` and in other cases we return `NextStep(leaf_opt)`.

- Declaration and side effect nodes:

```
186:       ty::TyAstNodeContent::SideEffect(_) => Ok(NodeConnection::NextStep(leaf_opt)),
```
(`leaf_opt` is initially `None`, and won't be `Some` until we've seen the first node)

This basically means that declaration and side effect nodes are skipped when constructing the linearized graph.

I can introduce a third variant to the `NodeConnection` enum, representing a node that we're skipping, if you prefer?

Reply:
I don't think that'll be necessary. This isn't exactly clean code but at least those reasons make sense.